jupyter_client/ioloop/manager.py

```diff
@@ -12,10 +16,51 @@
 from jupyter_client.manager import KernelManager
 
 
+class _ZMQStream(ZMQStream):
+    def _handle_recv(self):
+        """Handle a recv event."""
+        if self._flushed:
+            return
+        try:
+            msg = self.socket.recv_multipart(zmq.NOBLOCK, copy=self._recv_copy)
+            if isinstance(msg, asyncio.Future):
+                msg = msg.result()
+        except zmq.ZMQError as e:
+            if e.errno == zmq.EAGAIN:
+                # state changed since poll event
+                pass
+            else:
+                raise
+        else:
+            if self._recv_callback:
+                callback = self._recv_callback
+                self._run_callback(callback, msg)
+
+    def _handle_send(self):
+        """Handle a send event."""
+        if self._flushed:
+            return
+        if not self.sending():
+            gen_log.error("Shouldn't have handled a send event")
+            return
+
+        msg, kwargs = self._send_queue.get()
+        try:
+            status = self.socket.send_multipart(msg, **kwargs)
+            if isinstance(status, asyncio.Future):
+                status = status.result()
+        except zmq.ZMQError as e:
+            gen_log.error("SEND Error: %s", e)
+            status = e
+        if self._send_callback:
+            callback = self._send_callback
+            self._run_callback(callback, msg, status)
+
+
 def as_zmqstream(f):
     def wrapped(self, *args, **kwargs):
         socket = f(self, *args, **kwargs)
-        return ZMQStream(socket, self.loop)
+        return _ZMQStream(socket, self.loop)
```

Comment:
I've meant to make this easier/more public for ages, but this is the official way to instantiate a non-default socket type for a context, rather than needing a custom `_ZMQStream`:

```suggestion
        save_socket_type = None
        # zmqstreams only support sync sockets
        if self.context._socket_type is not zmq.Socket:
            save_socket_type = self.context._socket_type
            self.context._socket_type = zmq.Socket
        stream = ZMQStream(socket, self.loop)
        if save_socket_type:
            # restore default socket type
            self.context._socket_type = save_socket_type
        return stream
```

Reply:
(`with mock.patch.object` doesn't work because of a bug in `Context.__delattr__`)